agent/bench-scripts/pbench-specjbb2005

```diff
@@ -277,12 +277,12 @@ trap "interrupt" INT QUIT TERM
 
 # edit the jbb properties file to match user's options
 > ${benchmark_run_dir}/SPECjbb.props.sed
-printf "s/input.jvm_instances=.*/input.jvm_instances=$nr_jvms/\n" >> ${benchmark_run_dir}/SPECjbb.props.sed
-printf "s/input.measurement_seconds=.*/input.measurement_seconds=$runtime/\n" >> ${benchmark_run_dir}/SPECjbb.props.sed
-printf "s/input.starting_number_warehouses=.*/input.starting_number_warehouses=$start_warehouses/\n" >> ${benchmark_run_dir}/SPECjbb.props.sed
-printf "s/input.increment_number_warehouses=.*/input.increment_number_warehouses=$inc_warehouses/\n" >> ${benchmark_run_dir}/SPECjbb.props.sed
-printf "s/input.ending_number_warehouses=.*/input.ending_number_warehouses=$stop_warehouses/\n" >> ${benchmark_run_dir}/SPECjbb.props.sed
-sed -f ${benchmark_run_dir}/SPECjbb.props.sed ${specjbb2005_dir}/SPECjbb.props > ${benchmark_run_dir}/SPECjbb.props
+printf -- "s/#?input.jvm_instances=.*/input.jvm_instances=$nr_jvms/\n" >> ${benchmark_run_dir}/SPECjbb.props.sed
```

Comment:
OK, what's the logic behind "if [the file] has a commented out line we now uncomment it"? Doesn't that make it pointless to comment a props line in the first place? Why would someone do that, and why do we want to ignore it if they do?

Granted, handling of this template props file is extremely opinionated, e.g., the `.*` instead of `_replace_`; but this particular transform seems a bit odd, and I'd love to see some rationale recorded in a comment or at least in the git log.

Reply:
This only handles the case of a comment-introducer followed by no whitespace or any other character (e.g., it won't entirely uncomment a line prefixed with `##`)...is this a bug or a feature?

In any case, my reaction is similar to Dave's:  this idea seems highly questionable.

The clincher for me is this:  if there is _no entry_ for `input.jvm_instances` in the file, we **don't add one**...so why would we uncomment one that we find there?  ("Commented-out" should be the same as "not there" in functional terms, so the two cases **_should_** be handled similarly.)

Reply:
The first question we might want to answer before this line of discussion is: what is the purpose of the `pbench-specjbb2005` bench script?  For that matter, what is the purpose of any of the CLI interfaces located in `bench-scripts`?

Reply:
It's a fair question, but take care that it doesn't lead us into a minefield.  😆 

I think your contention is that, if the user specified something like the number of JVMs on the command line, then that should override whatever they (or someone) put in the configuration file.  I don't disagree with that position, but, if that's the case, then `pbench-specjbb2005` should have a provision for _adding_ the `input.jvm_instances=` line to the file if it's not there.  (And, if we are going to do something like that, perhaps we can just append all the options that we want to the bottom and they will supersede any specifications earlier in the file, and we won't need to do stuff like trying to uncomment or otherwise edit the existing lines.)

But, if the purpose of the script is to allow the user to run the workload despite how s/he has configured it, what's so bad about it poking holes in the firewall?  Doing that is just "enabling the workload to run", as evidenced by the fact that it won't run without doing it, and, surely the user _wants_ us to do that, because he told us to run the workload, no?  🤣

Reply:
> It's a fair question, but take care that it doesn't lead us into a minefield. 😆

Exactly.  Which is why this is a small step and we can take more steps later in a proper benchmark workload runner script.

> I think your contention is that, if the user specified something like the number of JVMs on the command line, then that should override whatever they (or someone) put in the configuration file. I don't disagree with that position, but, if that's the case, then `pbench-specjbb2005` should have a provision for _adding_ the `input.jvm_instances=` line to the file if it's not there. (And, if we are going to do something like that, perhaps we can just append all the options that we want to the bottom and they will supersede any specifications earlier in the file, and we won't need to do stuff like trying to uncomment or otherwise edit the existing lines.)

Yes, but I am trying not to make huge changes.  This change represents work that was tried out on a broken system that fixed the problem and worked.

Future PRs can either make an improvement to this code, or move the support to a more proper benchmark workload runner script.

> But, if the purpose of the script is to allow the user to run the workload despite how s/he has configured it, what's so bad about it poking holes in the firewall?

It is not quite that.  The user is asking the script to make that configuration happen.

Poking holes in a firewall is a much bigger problem with wider implications that are not on the same level as how a benchmark is run.

Gratuitous software installs and firewall hole poking is not a role for the Pbench Agent and must be handled deliberately by the person or people responsible for that server's configuration.

> Doing that is just "enabling the workload to run", as evidenced by the fact that it won't run without doing it, and, surely the user _wants_ us to do that, because he told us to run the workload, no? 🤣

Ah ... no.  That is a contortion of reality.  =)

Reply:
Sure, that all sounds like good future work.

This is a really simple change that measurably improves this, does not cause a problem, and we can still iterate going forward.

I really want to avoid trying to get a single PR "perfect", but consider iterating across multiple PRs over time.

Reply:
It causes a problem because it only addresses the issue of commented-out config lines _some_ of the time.

This isn't about trying to get the PR to be perfect -- it's about deciding whether accepting the PR is better than not accepting it.

At the moment, we can state that commenting out lines in the config file causes problems and that the user can address those problems by un-commenting-out those lines.  If we accept this change, then we'll be in a position of saying that, if the user has commented out the lines in just the right way, our script will uncomment them, but otherwise things are just as broken as they ever were.

The value of that amount of improvement doesn't seem, to me, to offset the potential for resulting confusion...especially if there are better solutions within reach.

Can you point me to instructions on how to run SpecJBB (in particular, what I need to install and set up)?  I'd like to try adding overrides to the end of the config file to see if that will "just work".

Reply:
> It causes a problem because it only addresses the issue of commented-out config lines _some_ of the time.

How is that causing a problem?  The current code in 0.69 and in 0.71 does the same thing.  So this change, while is not complete, is not causing a bigger problem.  It is just not a complete solution.  It is incremental.  It is useful as demonstrated by unblocking a user of the interface.

Why would we not take the incremental change and focus on fixing the long term in the "next framework" instead of investing in this framework?

> This isn't about trying to get the PR to be perfect -- it's about deciding whether accepting the PR is better than not accepting it.

What is the argument that this helps the user by not accepting it?  There is nothing else in development that does this fix, and not likely to happen for a while given all the other priorities the project has.

> At the moment, we can state that commenting out lines in the config file causes problems and that the user can address those problems by un-commenting-out those lines. If we accept this change, then we'll be in a position of saying that, if the user has commented out the lines in just the right way, our script will uncomment them, but otherwise things are just as broken as they ever were.

Sure, but it does solve one case and it was demonstrated to be useful.  What is wrong with a small incremental change?

> The value of that amount of improvement doesn't seem, to me, to offset the potential for resulting confusion...especially if there are better solutions within reach.

Where is the confusion?  This patch when applied "live" solved a problem and allowed the user to move forward.

Are we really focusing on a real problem or one that just might happen in the future?  Let's invest our effort in a longer term solution for workloads rather than trying to polish this interface further.

We are trying to get 0.71 out the door.  This whole discussion has now potentially delayed the release 4 days.

> Can you point me to instructions on how to run SpecJBB (in particular, what I need to install and set up)? I'd like to try adding overrides to the end of the config file to see if that will "just work".

I don't know where they live in the wild.

Reply:
Please see [portante/pbench #20](https://github.com/portante/pbench/pull/20) for a better solution.

Reply:
Thanks for the alternative solution!  Along with working unit tests, I hope it works.

Reply:
I'm not sure how we would unit test this, given that our existing unit tests stub-out the thing which actually reads the configuration file....  Got any thoughts on that?

Reply:
Let me amend...do we need anything beyond the improvements to the unit testing which this PR delivers?